crates/phactory/src/system/mod.rs

```diff
@@ -1034,7 +1234,114 @@ impl<Platform: pal::Platform> System<Platform> {
+    fn process_batch_rotate_master_key(
+        &mut self,
+        block: &mut BlockInfo,
+        origin: MessageOrigin,
+        event: BatchRotateMasterKeyEvent,
+    ) -> Result<(), TransactionError> {
+        if !origin.is_gatekeeper() {
+            error!("Invalid origin {:?} sent a {:?}", origin, event);
+            return Err(TransactionError::BadOrigin.into());
+        }
+        // check the event sender identity and signature to ensure it's not forged with a leaked master key and really from
+        // a gatekeeper
+        let data = event.data_be_signed();
+        let sig = sp_core::sr25519::Signature::try_from(event.sig.as_slice())
+            .or(Err(TransactionError::BadSenderSignature))?;
+        if !sp_io::crypto::sr25519_verify(&sig, &data, &event.sender) {
+            return Err(TransactionError::BadSenderSignature.into());
+        }
+        if !chain_state::is_gatekeeper(&event.sender, block.storage) {
+            error!("Fatal error: Forged batch master key rotation {:?}", event);
+            return Err(TransactionError::MasterKeyLeakage);
+        }
+
+        let my_pubkey = self.identity_key.public();
+        if event.secret_keys.contains_key(&my_pubkey) {
+            if self.gatekeeper.is_none() {
+                panic!(
+                    "Batch rotate master key to a normal worker {:?}",
+                    &my_pubkey
+                );
+            }
+
+            let encrypted_key = &event.secret_keys[&my_pubkey];
+            let new_master_key = self.decrypt_key_from(
+                &encrypted_key.ecdh_pubkey,
+                &encrypted_key.encrypted_key,
+                &encrypted_key.iv,
+            );
+
+            info!("Worker: successfully decrypt received rotated master key");
+            self.handle_new_master_key(new_master_key, event.rotation_id);
+        } else if self.gatekeeper.is_some() {
+            // will be here in two situations
+            if self.master_key_history.last().unwrap()
+                == &self.master_key.as_ref().unwrap().dump_secret_key()
+            {
+                // 1. This is an unregistered GK whose master key is not outdated yet.
+                // Delete the gatekeeper and master_key since it cannot do silent syncing anymore
+                info!("Worker: master key rotation receiced, stop gatekeeper silent syncing");
+                self.master_key_cleanup();
+            } else {
+                // 2. This is a valid GK in syncing, the needed master key should already be dispatched in the "future".
+                // Tick the master key correspondingly
+                let current_master_key = self.master_key.as_ref().unwrap().dump_secret_key();
+                let current_idx = self
+                    .master_key_history
+                    .iter()
+                    .position(|key| key == &current_master_key)
+                    .expect("Corrupted master key history");
```

Comment:
Looks better also storing the rotation_id in the history vec, and matching with it.

Reply:
Fixed.